### DISCUSSION TITLE
Edit Router operations

### DISCUSSION THREAD

**Discussion by gracekumagai** [Autodesk/maya-usd#2376] [Q&A]
I have been testing the [Edit Router](https://github.com/Autodesk/maya-usd/blob/3885eb01362ef751967e9ca200a304566f8ff723/test/lib/ufe/testDuplicateCmd.py#L361) as we are hoping to use it to route edits to the session layer that we don't want to pass downstream (visibility, pickability, etc).

Are there other operations that are supported (in addition to parenting and duplicating)? If so, I was wondering where I can find the list? I tried to pass "visibility" as the operation (`mayaUsd.lib.registerEditRouter("visibility", ...)`) and after making the prim invisible, the edit was not authored on the session layer.

Thank you so much!

**Comment by ppt-adsk** [comment]
Hi @gracekumagai, the currently-supported operations for edit routing are here:
https://github.com/Autodesk/maya-usd/blob/9f3800b1aab770293773c3146af975b4b3678e3a/lib/mayaUsd/utils/editRouter.cpp#L183
and here:
https://github.com/Autodesk/maya-usd/blob/9f3800b1aab770293773c3146af975b4b3678e3a/lib/mayaUsd/utils/editRouter.cpp#L188
So mayaReferencePush, parent and duplicate for now.  Once an edit router is registered, the edit routing registry must be accessed to retrieve the edit router and call it.  This requires changing the code that writes to the USD data model.  Here is the implementation for parent:
https://github.com/Autodesk/maya-usd/blob/9f3800b1aab770293773c3146af975b4b3678e3a/lib/mayaUsd/ufe/UsdUndoInsertChildCommand.cpp#L96
For duplicate:
https://github.com/Autodesk/maya-usd/blob/9f3800b1aab770293773c3146af975b4b3678e3a/lib/mayaUsd/ufe/UsdUndoDuplicateCommand.cpp#L54
And for mayaReferencePush (more complicated example):
https://github.com/Autodesk/maya-usd/blob/8d0a42532bbdc8294d691270811f8b3ebf42a258/lib/usd/translators/mayaReferenceUpdater.cpp#L195
I believe that to implement edit routing for visibility, you would need to change the UsdUndoVisibleCommand, somewhere here:
https://github.com/Autodesk/maya-usd/blob/600c31abcd68ef3fe0047b28fb5f33f6edd0872f/lib/mayaUsd/ufe/UsdUndoVisibleCommand.cpp#L55
You would save the current edit target, set it to what your edit router returns, then set it back to its previous value.  If the edit target layer is not prim-dependent, you can call upon the edit router in the UsdUndoVisibleCommand() constructor, and further if it's a constant (e.g. the session layer) ask the edit router once and save the result.  Hope this helps, cheers.

**Reply by dj-mcg** [reply]
Thanks @ppt-adsk - were there plans on your end to continue broadening the support for operations that could be routed in this manner? If there were plans to add support for visibility on your end I'd prefer prioritizing the other tasks we know we need to tackle ourselves before circling back to this one. Thanks so much!

**Comment by dj-mcg** [comment]
Just to followup - it sounds like adding routing operations is something we could implement and contribute, which would allow the mayaUsd team to focus on new features and fixes that might not be plugin-only, so we are happy to schedule this work on our end. Thanks @ppt-adsk  and others!

**Reply by ppt-adsk** [reply]
Hi @dj-mcg, took at crack at it just to test out that it wouldn't be too bad to implement support for this.  Have a look at
https://github.com/Autodesk/maya-usd/pull/2438
Writing visibility edits to the session layer is done in the unit test, so this should be what you're looking for, unless I'm mistaken.  Let me know what you think!  Cheers.

**Comment by dj-mcg** [comment]
Rad! Thanks for trying it out and sharing your example!
